src/api/routes/__tests/groups.route.spec.ts

```diff
@@ -0,0 +1,162 @@
+import express, { json } from 'express';
+import request from 'supertest';
+
+import groups from '../groups.route';
+
+const app = express();
+app.use(json());
+app.use('/', groups);
+
+const group = {
+    id: '1',
+    name: 'Admin',
+    permissions: ['READ', 'WRITE', 'DELETE', 'SHARE', 'UPLOAD FILES']
+};
+
+const createdGroup = {
+    name: 'User',
+    permissions: ['READ', 'WRITE']
+};
+
+const updatedGroup = {
+    id: '1',
+    name: 'Updated Group',
+    permissions: ['READ', 'WRITE', 'DELETE', 'SHARE', 'UPLOAD FILES']
+};
+
+const userToAdd = {
+    id: '1',
+    login: 'Test User',
+    password: 'Testpass1234',
+    age: 23,
+    isDeleted: false
+};
+
+jest.mock('../../../middlewares/auth/secureRoute.auth.ts', () =>
+    jest.fn((req, res, next) => next())
+);
+
+jest.mock('../../../services/group.service.ts', () => {
```

Comment:
This would be a far better way of approaching this (Then there's only be 2/3 tests files needed). I've been having some trouble trying to mocking the model dependencies though =/. I've tried to use spies on the route tests (e.g. `jest.spyOn(UserModel, 'findAll').mockReturnValue(Promise.resolve([user1]));`) and mocking the return value, but it either returns a function with these methods missing or returns a mock constructor - both indicating the instance methods are undefined.

Reply:
What about `jest.spyOn(UserModel, "findAll").mockImplementation(() => Promise.resolve(...));`?

Reply:
I totally agree. I was trying to find a way to not relying on the order of the calls when mocking the service, but I hit a brick wall. 

Fortunately I think I found a good way to mock the return value in each case: 

```
const mockFindAll: jest.Mock = jest.fn();

jest.mock('../../../models', () => {
    return {
        User: {
            findAll: () => mockFindAll(),
        }
    };
})
```

```
    test("'get' method for '/users/' is called and returns the correct response", async () => {
        mockFindAll.mockImplementationOnce(() => Promise.resolve([user1]));

        const res = await request(app).get('/');

        expect(res.statusCode).toBe(200);
        expect(res.body).toEqual([user1]);
    });

```